finat/point_set.py

```diff
@@ -34,15 +34,23 @@ def expression(self):
 
 
 class PointSingleton(AbstractPointSet):
-    """Just a single point."""
+    """A point set representing a single point. It therefore has no indices an
+    a `gem.Literal` expression."""
 
     def __init__(self, point):
+        """Build a PointSingleton from a single point.
+
+        :arg point: A single point which, once turned into a numpy array, has
+            shape (point dimension,)."""
         point = numpy.asarray(point)
+        # Ensure we only have 1 point
         assert len(point.shape) == 1
```

Comment:
The assertion is a bit stronger than the comment: the point set shape is basically scalar `()`, rather than a single- element vector `(1,)` or 1 x 1 matrix `(1, 1)`.  So `point.shape` is merely the shape of the point, i.e. a vector.

Reply:
I'm not sure I follow - the `point` has shape `(point dimension,)` not `()` here and the assertion checks that no?

Reply:
Updated comment

Reply:
(The thing that I found contentious was that I couldn't initialise a PointSingleton with something of shape `(point dimension, 1)` and have it just work - you have to squeeze down the number of dimensions such that you have something of shape `(point dimension, )`)

Reply:
The shape of the internal array is always `point_set_shape + (point_dimension,)` in the general case, and in case of `PointSingleton` `point_set_shape == ()`, while in case of `PointSet` `point_set_shape = (npoints,)`.

Reply:
Okay I've tried to make clarify this in 2066f381b157db00075124e58626c83c00fed7e1 . A `PointSet` deals with a vector of points specifically whilst a `PointSingleton` is a single scalar point.

I suppose it might be nice one day to be able to make an `NxMxL` `TensorPointSet` of points with dimension `D` directly from a numpy `ndarray` of shape `(N, M, L) + (D) = (N, M, L, D)`

Reply:
Sure, although normally you're carrying around pointsets by this point, no?

Reply:
Yes, normally you wouldn't want to construct the outer product ndarray of a TensorPointSet, but rather just tabulate the factor elements of a tensor product element on the factor point sets of the tensor point set...